docs/architecture/adr-015-ibc-packet-receiver.md

```diff
@@ -0,0 +1,289 @@
+  result = app.runMsgs(runMsgCtx, msgs)
+  // BEGIN modification made in this ADR
+  if result.IsOK() || result.IsBreak() {
+  // END
+    msCache.Write()
+  }
+
+  return result
+}
+```
+
+The Cosmos SDK will define an `AnteDecorator` for IBC packet receiving. The
+`AnteDecorator` will iterate over the messages included in the transaction, type
+`switch` to check whether the message contains an incoming IBC packet, and if so
+verify the Merkle proof.
+
+```go
+type ProofVerificationDecorator struct {
+  clientKeeper ClientKeeper
+  channelKeeper ChannelKeeper
+}
+
+func (pvr ProofVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+  for _, msg := range tx.GetMsgs() {
+    var err error
+    switch msg := msg.(type) {
+    case client.MsgUpdateClient:
+      err = pvr.clientKeeper.UpdateClient(msg.ClientID, msg.Header)
+    case channel.MsgPacket:
+      err = pvr.channelKeeper.RecvPacket(msg.Packet, msg.Proofs, msg.ProofHeight)
+    case chanel.MsgAcknowledgement:
+      err = pvr.channelKeeper.AcknowledgementPacket(msg.Acknowledgement, msg.Proof, msg.ProofHeight)
+    case channel.MsgTimeoutPacket:
+      err = pvr.channelKeeper.TimeoutPacket(msg.Packet, msg.Proof, msg.ProofHeight, msg.NextSequenceRecv)
+    case channel.MsgChannelOpenInit;
+      err = pvr.channelKeeper.CheckOpen(msg.PortID, msg.ChannelID, msg.Channel)
+    default:
+      continue
+    }
+
+    if err != nil {
+      return ctx, err
+    }
+  }
+  
+  return next(ctx, tx, simulate)
+}
+```
+
+Where `MsgUpdateClient`, `MsgPacket`, `MsgAcknowledgement`, `MsgTimeoutPacket`
+are `sdk.Msg` types correspond to `handleUpdateClient`, `handleRecvPacket`,
+`handleAcknowledgementPacket`, `handleTimeoutPacket` of the routing module,
+respectively.
+
+The side effects of `RecvPacket`, `VerifyAcknowledgement`, 
+`VerifyTimeout` will be extracted out into separated functions, which will
+be called by the application handlers after the execution.
+
+```go
+func (keeper ChannelKeeper) WriteAcknowledgement(ctx sdk.Context, packet Packet, ack []byte) {
+  keeper.SetPacketAcknowledgement(ctx, packet.GetDestPort(), packet.GetDestChannel(), packet.GetSequence(), ack)
+  keeper.SetNextSequenceRecv(ctx, packet.GetDestPort(), packet.GetDestChannel(), packet.GetSequence())
+}
+
+func (keeper ChannelKeeper) DeleteCommitment(ctx sdk.Context, packet Packet) {
+  keeper.deletePacketCommitment(ctx, packet.GetSourcePort(), packet.GetSourceChannel(), packet.GetSequence())
+}
+
+func (keeper ChannelKeeper) DeleteCommitmentTimeout(ctx sdk.Context, packet Packet) {
+  k.deletePacketCommitment(ctx, packet.GetSourcePort(), packet.GetSourceChannel(), packet.GetSequence())
+  
+  if channel.Ordering == types.ORDERED [
+    channel.State = types.CLOSED
+    k.SetChannel(ctx, packet.GetSourcePort(), packet.GetSourceChannel(), channel)
+  }
+}
+```
+
+Each application handler should call respective finalization methods on the `PortKeeper`
```

Comment:
Will the finalization methods have to be called by each module handler or do we want to handle it automatically?

Reply:
Finalization methods have to be called by each module handler.

Reply:
>Finalization methods have to be called by each module handler.

please write that down in the ADR